### ISSUE TITLE
[Bug]: Kernel Oops in BPF netdata_tcp_sendmsg

### ISSUE THREAD

**Issue by davidmehren** [netdata/netdata#12027]
### Bug description

All hosts on our Proxmox cluster run Netdata. Yesterday, we encountered a repeating kernel crash on two nodes, which appear to have been caused by Netdatas eBPF plugin (at least "netdata" appears in the call trace). That crash did not take the host down directly, but caused Ceph to stop answering I/O requests, leading to outages for all services on the host. The hosts also did not reboot cleanly, but required a hard reset.

The traces are all somewhat similar, one example (register values removed):
```
BUG: unable to handle page fault for address: ffffffffff6004ad
#PF: supervisor read access in kernel mode
#PF: error_code(0x0000) - not-present page
PGD cbc215067 P4D cbc215067 PUD cbc217067 PMD cbc219067 PTE 0
Oops: 0000 [#2] SMP NOPTI
CPU: 60 PID: 3174 Comm: msgr-worker-1 Tainted: P      D    O      5.15.12-1-pve #1
Hardware name: Supermicro Super Server/H11DSi-NT, BIOS 1.3 06/25/2019
RIP: 0010:copy_from_kernel_nofault+0xa7/0x100
<snip>
Call Trace:
 <TASK>
 bpf_probe_read_compat+0x64/0x90
 bpf_prog_896bb3a8c2cab17b_netdata_tcp_sendmsg+0xbc/0xb48
 ? __mod_timer+0x271/0x430
 trace_call_bpf+0x9f/0x150
 kretprobe_perf_func+0x4e/0x280
 ? tcp_schedule_loss_probe.part.0+0x13c/0x1a0
 ? tcp_write_xmit+0x98e/0x1480
 kretprobe_dispatcher+0x6c/0x80
 ? inet_sendmsg+0x43/0x70
 __kretprobe_trampoline_handler+0xb3/0xe0
 trampoline_handler+0x48/0x60
 kretprobe_trampoline+0x2a/0x50
RIP: 0010:kretprobe_trampoline+0x0/0x50
<snip>
 sock_sendmsg+0x5e/0x70
 ____sys_sendmsg+0x257/0x2a0
 ? import_iovec+0x31/0x40
 ? sendmsg_copy_msghdr+0x7e/0xa0
 ___sys_sendmsg+0x82/0xc0
 ? bpf_ktime_get_ns+0xe/0x10
 ? __fget_files+0xa3/0xd0
 ? __fget_light+0x32/0x80
 __sys_sendmsg+0x62/0xb0
 ? __secure_computing+0x43/0xe0
 __x64_sys_sendmsg+0x1f/0x30
 do_syscall_64+0x5c/0xc0
 ? irqentry_exit_to_user_mode+0x9/0x20
 ? irqentry_exit+0x19/0x30
 ? common_interrupt+0x55/0xa0
 ? asm_common_interrupt+0x8/0x40
 entry_SYSCALL_64_after_hwframe+0x44/0xae
RIP: 0033:0x7f7c56507e4d
<snip>
 </TASK>
Modules linked in: xt_mac xt_nat xfrm_user overlay binfmt_misc veth rbd ceph libceph fscache netfs ebtable_filter ebtables ip6table_raw ip6t_REJECT nf_reject_ipv6 ip6table_filter ip6_tables iptable_raw ipt_REJECT nf_reject_i>
 zzstd(O) zlua(O) zavl(PO) icp(PO) zcommon(PO) znvpair(PO) spl(O) vhost_net vhost vhost_iotlb tap ib_iser rdma_cm iw_cm ib_cm ib_core iscsi_tcp libiscsi_tcp libiscsi scsi_transport_iscsi drm sunrpc ip_tables x_tables autofs4>
CR2: ffffffffff6004ad
---[ end trace 22a4446218922888 ]---
RIP: 0010:copy_from_kernel_nofault+0xa7/0x100
<snip>
```
The call traces always have `RIP: 0010:copy_from_kernel_nofault+0xa7/0x100`, `bpf_probe_read_compat+0x64/0x90` and `bpf_prog_896bb3a8c2cab17b_netdata_tcp_sendmsg+0xbc/0xde8` at the top, but differ slightly after that. If required, I can post them all in full.

In total, we had six such crashes, which reoccurred after a reboot, until we disabled the ebpf plugin.

The hosts ran kernel `5.15.7-1-pve` or `5.15.12-1-pve`, I think the Proxmox kernels are based on Ubuntu.

### Expected behavior

No kernel oops occurs.

### Steps to reproduce

I'm really not sure if this is reproducible, it happened spontaneously on two of our three hosts. It reoccured every one to two hours, until we disabled the ebpf plugin.

### Installation method

native binary packages (.deb/.rpm)

### System info

```shell
Linux legion 5.15.12-1-pve #1 SMP PVE 5.15.12-2 (Tue, 11 Jan 2022 16:43:18 +0100) x86_64 GNU/Linux
/etc/os-release:PRETTY_NAME="Debian GNU/Linux 11 (bullseye)"
/etc/os-release:NAME="Debian GNU/Linux"
/etc/os-release:VERSION_ID="11"
/etc/os-release:VERSION="11 (bullseye)"
/etc/os-release:VERSION_CODENAME=bullseye
/etc/os-release:ID=debian
```

### Netdata build info

```shell
Version: netdata v1.32.1-111-gaec2911bf
Configure options:  '--build=x86_64-linux-gnu' '--includedir=${prefix}/include' '--mandir=${prefix}/share/man' '--infodir=${prefix}/share/info' '--disable-option-checking' '--disable-silent-rules' '--libdir=${prefix}/lib/x86_64-linux-gnu' '--libexecdir=${prefix}/lib/x86_64-linux-gnu' '--disable-maintainer-mode' '--disable-dependency-tracking' '--prefix=/usr' '--sysconfdir=/etc' '--localstatedir=/var' '--libdir=/usr/lib' '--libexecdir=/usr/libexec' '--with-user=netdata' '--with-math' '--with-zlib' '--with-webdir=/var/lib/netdata/www' 'build_alias=x86_64-linux-gnu' 'CFLAGS=-g -O2 -ffile-prefix-map=/usr/src/netdata=. -fstack-protector-strong -Wformat -Werror=format-security' 'LDFLAGS=-Wl,-z,relro' 'CPPFLAGS=-Wdate-time -D_FORTIFY_SOURCE=2' 'CXXFLAGS=-g -O2 -ffile-prefix-map=/usr/src/netdata=. -fstack-protector-strong -Wformat -Werror=format-security'
Features:
    dbengine:                   YES
    Native HTTPS:               YES
    Netdata Cloud:              YES 
    ACLK Next Generation:       YES
    ACLK-NG New Cloud Protocol: YES
    ACLK Legacy:                NO
    TLS Host Verification:      YES
    Machine Learning:           YES
    Stream Compression:         YES
Libraries:
    protobuf:                YES (system)
    jemalloc:                NO
    JSON-C:                  YES
    libcap:                  NO
    libcrypto:               YES
    libm:                    YES
    tcalloc:                 NO
    zlib:                    YES
Plugins:
    apps:                    YES
    cgroup Network Tracking: YES
    CUPS:                    YES
    EBPF:                    YES
    IPMI:                    YES
    NFACCT:                  YES
    perf:                    YES
    slabinfo:                YES
    Xen:                     NO
    Xen VBD Error Tracking:  NO
Exporters:
    AWS Kinesis:             NO
    GCP PubSub:              NO
    MongoDB:                 NO
    Prometheus Remote Write: YES
```

**Comment by thiagoftsm** [comment]
Hello @davidmehren ,

We already observed problems on Ubuntu kernels, this happens because they do not use the exactly same kernel delivered by Linux Kernel team (https://www.kernel.org). An eBPF programs runs inside an environment that  is like a Virtual Machine, so it cannot crash a host.

To solve the problems we are having with Ubuntu's kernel, we created a file where we add kernels that create problems:

```conf
/etc/netdata/ebpf.d/ebpf_kernel_reject_list.txt
```

As you can see, only Ubuntu is listed inside it.

If it is necessary you continue running eBPF, you can disable only the socket editing `/etc/netdata/ebpf.d.conf` and set:

```conf
[ebpf programs]
     socket =  no
```

This will disable only the `socket` thread.

It is possible that when you have another kernel update the problem will be fixed.

I have used `5.15.x` since it was released on Slackware current and on Arch, but this last changed for `5.16.x` recently. I never observed the problem that you reported, so this can confirm that we are talking about a possible kernel with patches applied wrong.

Finally I won't consider this answer as final word, I will install the same Proxmox that you have, and I will try to replicate the error.

**Comment by thiagoftsm** [comment]
Hello @davidmehren ,

I ran some tests with `Proxmox VE 7.1-2`, and all the tests were successful as you can see in next logs:

| Log | Description |
|-------|----------------|
| [error.txt](https://github.com/netdata/netdata/files/7918637/error.txt) | Error log copied from `/var/log/netdata/error.log` |
|  [map_show.txt](https://github.com/netdata/netdata/files/7918639/map_show.txt) | Output of the command `bpftool map show` |
| [dump_tbl_global_sock.txt](https://github.com/netdata/netdata/files/7918640/dump_tbl_global_sock.txt) | Output of the command ` bpftool map dump name tbl_global_sock `. This is the global data. |
| [dump_tbl_bandwidth.txt](https://github.com/netdata/netdata/files/7918644/dump_tbl_bandwidth.txt) | Output of the command `bpftool map dump name tbl_bandwidth`. This is socket information per PID. |

I ran the following tests:

1. I installed Proxmox from scratch on Virtual Box
2. I updated the dist running the commands:
```sh
# apt-get update
# apt-get upgrade
# apt-get dist-upgrade
```
3. I installed netdata using a development branch that has everything already merged in our master branch and additional charts that we will bring during this quarter. I ran the following commands to install:
```sh
# ./packaging/installer/install-required-packages.sh --dont-wait
# ./netdata-installer.sh --dont-wait 
```
4. I ran netdata running more than 3 hours. During this time, I wrote a script to get files from different web servers in my internal network.

I did not observe any problem while I was trying to debug, but it is necessary to say something very important. I observed that after to install and upgrade the packages, I never received a kernel with the version `5.15.x`, instead, the official package for kernel is:

```sh
# uname -a
Linux test 5.13.19-2-pve #1 SMP PVE 5.13.19-4 (Mon, 29 Nov 2021 12:10:09 +0100) x86_64 GNU/Linux
 # dpkg -l | grep pve-kernel
ii  pve-firmware                         3.3-3                          all          Binary firmware code for the pve-kernel
ii  pve-kernel-5.13                      7.1-5                          all          Latest Proxmox VE Kernel Image
ii  pve-kernel-5.13.19-2-pve             5.13.19-4                      amd64        The Proxmox PVE Kernel Image
ii  pve-kernel-helper                    7.1-6                          all          Function for various kernel maintenance tasks.
```

How did you install the kernel `5.15.x` on your Proxmox?

Best regards,

Thiago

**Comment by davidmehren** [comment]
Thanks, @thiagoftsm, for taking the time to debug this!

Kernel 5.15 is currently optional in PVE, you can install it with `apt install pve-kernel-5.15`

**Comment by ThomasLamprecht** [comment]
> We already observed problems on Ubuntu kernels, this happens because they do not use the exactly same kernel delivered by Linux Kernel team (https://www.kernel.org). An eBPF programs runs inside an environment that is like a Virtual Machine, so it cannot crash a host.

I mean that's the case with basically every distro, there'll be a few patches on top and there'll also be a distro specific KConfig.

But, FWICT the bpf stack isn't (custom) patched by either Ubuntu nor us (for the latter I'm 100% sure) at all, and the eBPF VM or its verifier has not been exactly bug free in the past, so a regression or altered behavior could cause this.

> To solve the problems we are having with Ubuntu's kernel, we created a file where we add kernels that create problems:

Did you actually reach out to Ubuntu or looked more closely into this? At least I did not find any message on their kernel list.
Because if there's just a missing or "wrong" KConfig set, then reaching out to the Ubuntu kernel team (via mailing list or launchpad) may help to address your friction with their kernel relatively fast, as of now they may not even be aware of an issue with your tooling.

Apart from that I agree though, a user space eBPF program should never be able to trigger a kernel oops like that, at least if netdata does not do any other weird stuff (loading their own modules or such).

@davidmehren If you do not use ZFS and spare the time and a host for experimenting, you could try the mainline build from Ubuntu's mainline ppa, maybe it helps to ballpark the issue:

E.g., the equivalent of the kernel you use now (15.15.12) https://kernel.ubuntu.com/~kernel-ppa/mainline/v5.15.12/

Maybe also the most recent of the 5.15.x stable tree (even if I did not saw any really obvious fix for this in the git log between 5.15.12 and 5.15.16): https://kernel.ubuntu.com/~kernel-ppa/mainline/v5.15.16/

Depending on above results, a newer major release could be interesting too, e.g.: https://kernel.ubuntu.com/~kernel-ppa/mainline/v5.16.2/

**Comment by thiagoftsm** [comment]
Hello @davidmehren ,

I installed the kernel `5.15.12-1-pve` today, but before to install it, I had to add a Test repo, the repos I was using did not have the package. After this I restart the VM and I ran  the tests using kernel `Linux test 5.15.12-1-pve`:

```sh
# dpkg -l | grep pve-kernel
ii  pve-firmware                         3.3-3                          all          Binary firmware code for the pve-kernel
ii  pve-kernel-5.13                      7.1-5                          all          Latest Proxmox VE Kernel Image
ii  pve-kernel-5.13.19-2-pve             5.13.19-4                      amd64        The Proxmox PVE Kernel Image
ii  pve-kernel-5.15                      7.1-8                          all          Latest Proxmox VE Kernel Image
ii  pve-kernel-5.15.12-1-pve             5.15.12-3                      amd64        The Proxmox PVE Kernel Image
ii  pve-kernel-helper                    7.1-6                          all          Function for various kernel maintenance tasks.
# uname -a
Linux test 5.15.12-1-pve #1 SMP PVE 5.15.12-3 (Thu, 20 Jan 2022 16:36:44 +0100) x86_64 GNU/Linux
```

After to install the package on my VM, I ran netdata again during [3 hours](https://github.com/netdata/netdata/files/7921832/error.txt). The data were corrected stored in two hash tables ([App](https://github.com/netdata/netdata/files/7921835/dump_tbl_bandwidth.txt) and 
[Global](https://github.com/netdata/netdata/files/7921836/dump_tbl_global_sock.txt). And my VM did not crash any time. 

I agree with words given by @ThomasLamprecht , and I have a question to you, do you compile your kernel to change any configuration? Or do you use the official release? I am asking this, because your `uname` output is different of mine.

**Comment by thiagoftsm** [comment]
Hello @ThomasLamprecht ,

Firstly thank you very much for your collaboration!

> But, FWICT the bpf stack isn't (custom) patched by either Ubuntu nor us (for the latter I'm 100% sure) at all, and the eBPF VM or its verifier has not been exactly bug free in the past, so a regression or altered behavior could cause this.

I did not know these details, because Debian family is not my preferential distributions. I apologize for wrong words I used.

> Did you actually reach out to Ubuntu or looked more closely into this? 

When user reported for us the problem, in the next day there were a kernel update that fixed everything. We disabled Ubuntu after this, because we received warnings from another user having exactly the same problem with another update. The Ubuntu that we received reports was `18.04` running kernel `4.18.x`. On the same distribution version, we never detected problems neither with kernel `4.15.x` nor with kernel `5.4.x`.

> Apart from that I agree though, a user space eBPF program should never be able to trigger a kernel oops like that, at least if netdata does not do any other weird stuff (loading their own modules or such).

We do not install kernel modules, the motive that we decided to bring `eBPF` were the VMs they have. It is not our intention to create modules due their permissions.

Finally, today I recompiled from scratch the kernel `5.15.16` today and I also could run `eBPF.plugin` without any crash, but I did not use Proxmox, I used `Slackware current`

**Comment by ThomasLamprecht** [comment]
> but before to install it, I had to add a Test repo, the repos I was using did not have the package

FYI, if you freshly installed and have no subscription you'd need to add *some* Proxmox VE repo, as the default enterprise repo is only accessible with a subscription, `pvetest` naturally does the trick, but the more stable `pve-no-subscription` that most people use without subscription works too. https://pve.proxmox.com/wiki/Package_Repositories

> After to install the package on my VM, I ran netdata again during 3 hours. The data were corrected stored in two hash tables (App and Global. And my VM did not crash any **time.**

Great to hear, thanks for investing time into testing this with PVE directly!

 > The Ubuntu that we received reports was 18.04 running kernel 4.18.x.

Hmm, that's a bit odd as the Ubuntu LTS 18.04 release is actually using 4.15 as their recommended default, the bionic kernel repo's main branch also still points at 4.15 https://kernel.ubuntu.com/git/ubuntu/ubuntu-bionic.git/log/
So it could be that this user installed either some other kernel offering that was missing bpf support or something the like.

> We do not install kernel modules, the motive that we decided to bring eBPF were the VMs they have. It is not our intention to create modules due their permissions.

OK, thanks for your clarification.

> Finally, today I recompiled from scratch the kernel 5.15.16 today and I also could run eBPF.plugin without any crash, but I did not use Proxmox, I used Slackware current.

Good to know that there's no general issue on the latest 5.15.x tag too.

So, this could be a HW specific issue for @davidmehren, and thus I'd like to wait on @davidmehren test feedback on this.
FWIW, we have the same/similar mainboard in our testlab IIRC `Supermicro Super Server/H11DSi-NT` maybe I can get around to run some plain test there too.

**Comment by thiagoftsm** [comment]
Hello @ThomasLamprecht ,

I am glad I could help you! Now that I have a VM with Proxmox, and people like @davidmehren is using `eBPF.plugin` on it, I will continue testing on this distribution to be sure that everything is going fine.

> Hmm, that's a bit odd as the Ubuntu LTS 18.04 release is actually using 4.15 as their recommended default, the bionic kernel repo's main branch also still points at 4.15 https://kernel.ubuntu.com/git/ubuntu/ubuntu-bionic.git/log/

Now it is clear what was happening on Ubuntu. I always observed the `4.15` as a very stable kernel on it.

@davidmehren let us know if you need any help.

**Comment by davidmehren** [comment]
Thanks for your help, everyone.

We will try switching to the 5.15 mainline kernel on one of the affected hosts on the next weekend, and then re-enable the eBPF plugin in Netdata.

I fear the bug might only appear in specific setups, possibly only when Ceph is used. When the bug appeared on our hosts, it caused all Ceph I/O to stop in 6 out of 6 cases, but seemingly affected no other network traffic. One of the traces even contains `Workqueue: ceph-msgr ceph_con_workfn [libceph]`.

**Comment by thiagoftsm** [comment]
Hello @davidmehren ,

It sounds a good plan, we will wait for your update.
I agree with you that this looks like a specific setup.

Best regards!

**Comment by davidmehren** [comment]
After running with the Ubuntu mainline kernel `5.15.12-051512-generic` for about a day, we had another kernel oops.

I'm now pretty sure it is Ceph-related. My theory is this:
- The eBPF VM crashes, taking out a network thread (not sure if "thread" is the correct term here) of a Ceph OSD.
- The affected OSD now cannot transfer data over the network anymore, but does not realize that. Errors like `osd.XX 9562 heartbeat_check: no reply from <IP>:6838 osd.YY since ...` appear in the journals of the hosts in the Ceph cluster.
- For unknown reasons, the affected OSD is often not marked `out` by the cluster, so the cluster waits for IO on the OSD to complete. As the OSD has non-working networking, this will lead nowhere.
- `iowait` starts to accumulate on hosts accessing the Ceph cluster.
- Manually restarting the affected OSD (`systemctl restart ceph-osd@XX`) immediately fixes the problem and the cluster starts recovering.

So, where are we now?
I'm not sure if it is Netdata's fault for running eBPF code that crashes the kernel, or the kernel's fault for being crashed by the eBPF VM.

Unfortunately, I will not be able to test more kernels, as users will start to yell at me if I continue to impact the production cluster 😅
Maybe @ThomasLamprecht at Proxmox has a Ceph cluster for testing and can try to reproduce the issue?

**Comment by thiagoftsm** [comment]
Hello @davidmehren ,

Thank you for your additional report. 

Right now the eBPF.plugin is only attaching `kprobe` to few socket functions, but this will change in a near future, because we will work with `trampoline` when it is available. The simple fact to add `kprobes` for a specific function should not break the whole environment considering the expected behavior for eBPF programs. Considering everything you described, I have two questions for you:

1. How much memory do you have for each Ceph host?
2. Can you attach your `/proc/config.gz` for we verify how your kernel was compiled?

I think there is a small probability that our plugin is crashing the host, because this is not expected from eBPF, but we will try to help you the max we can.

Best regards!

**Comment by davidmehren** [comment]
@thiagoftsm 
Two nodes have 256 GB memory, one has 128.
This is the kernel config of the Proxmox kernel:
[config-5.15.12-1-pve.txt](https://github.com/netdata/netdata/files/7987802/config-5.15.12-1-pve.txt)

**Comment by vobruba-martin** [comment]
@thiagoftsm 
My server has 256GB memory
Ubuntu kernel config: [config-5.13.0-27-generic.ubuntu-20.04.txt](https://github.com/netdata/netdata/files/7987857/config-5.13.0-27-generic.ubuntu-20.04.txt)

**Comment by vobruba-martin** [comment]
Running 24 hours without any issue. Looks like `socket = no` is a good workaround for now.

**Comment by thiagoftsm** [comment]
Hello @vobruba-martin and @davidmehren ,

Firstly I am sorry for the delay to answer you, but yesterday I was returning from my vacation and I could not answer you.
Thank you very much to send me your config files for me to compile the kernel with the same configuration.

@vobruba-martin  I observed that when I installed the Ubuntu with image from this [link](https://releases.ubuntu.com/20.04.3/ubuntu-20.04.3-live-server-amd64.iso?_ga=2.170294176.1010961149.1643893380-2091835765.1643030800), Ubuntu was installed with kernel `5.4.X`, so I had to install the kernel `5.13.0-27` manually. Am I missing something? I also observed that Ubuntu has reported security problems with some packages for version [5.13.x](https://ubuntu.com/security/notices/USN-5219-1), including the `5.13.25`  reported [here](https://github.com/netdata/netdata/issues/12073). On the other hand I suppose you updated your kernel since your reported the bug, because you gave me a config for `5.13.27`.

For everybody know, right now I am trying to recreate the problem with Ubuntu, and I am working to confirm what is creating the problems for you. I will be back with more information as soon I have more details.

Best regards,

Thiago

**Comment by davidmehren** [comment]
@thiagoftsm No problem, thanks for investigating this!

Kernel 5.13 is the current Hardware Enablement (HWE) kernel for Ubuntu 20.04, see https://packages.ubuntu.com/focal/linux-generic-hwe-20.04.
Ubuntu ships updated kernels for their LTS versions to support newer hardware, see https://ubuntu.com/about/release-cycle#ubuntu-kernel-release-cycle

If I understand correctly, a HWE kernel will be used on desktop installs automatically and can be manually installed using the package above in server installs.

**Comment by vobruba-martin** [comment]
@thiagoftsm Yes, I've updated the kernel when I was trying to solve the issue. I'm using `linux-image-generic-hwe-20.04-edge` package because of HW support. Versions I've tested are `5.13.0-25`/`5.13.0-27`, not `5.13.25`/`5.13.27`. Configs for `5.13.0-25` and `5.13.0-27` are the same.

```diff
$ diff /boot/config-5.13.0-25-generic /boot/config-5.13.0-27-generic 
3c3
< # Linux/x86 5.13.0-25-generic Kernel Configuration
---
> # Linux/x86 5.13.0-27-generic Kernel Configuration
46c46
< CONFIG_VERSION_SIGNATURE="Ubuntu 5.13.0-25.26~20.04.1-generic 5.13.19"
---
> CONFIG_VERSION_SIGNATURE="Ubuntu 5.13.0-27.29~20.04.1-generic 5.13.19"

```

**Comment by vobruba-martin** [comment]
@thiagoftsm I don't think it's important but the system was installed from the [MinimalCD image](http://archive.ubuntu.com/ubuntu/dists/focal/main/installer-amd64/current/legacy-images/netboot/mini.iso) and after that I installed `linux-image-generic-hwe-20.04-edge` to use the latest available HWE kernel.

**Comment by thiagoftsm** [comment]
Thank you for your answers I will return with more details about tests later today. :handshake:

**Comment by thiagoftsm** [comment]
Hello guys @vobruba-martin and @davidmehren , 

I was the whole day working and debugging and I observed that sometimes a race condition was happening, I made a PR disabling the thread to avoid future problems, but this won't stay forever, because I am already working to fix this and I expect to have another PR removing this condition in the next days.

Thank you very much to help us to debug this.

Best regards,

Thiago

**Comment by thiagoftsm** [comment]
Hello @vobruba-martin and @davidmehren,

Sorry for the delay, but I was waiting for netdata `1.33.1` to be released for me to bring the fix. I made the PR https://github.com/netdata/netdata/pull/12190 that is bringing the latest socket code I already merged with PR https://github.com/netdata/kernel-collector/pull/291.  While I was developing this new PR I did not observe any problem, including when I was using kernel `4.18` on Ubuntu `18.04` that always gave me problems, if you have a time, please, take a look in the PR description about how to test it, and give us your results.

Best regards!

**Comment by vobruba-martin** [comment]
Hi @thiagoftsm,

I always use kickstart script to install netdata so I'm not familiar with manual installation. Is it enough to clone the branch, run `cmake -DNETDATA_INTERNAL_CHECKS=1` and that's it? Files in `/etc/netdata` and `cache directory` directories stay intact and after testing I can safely revert to the current version?

**Comment by thiagoftsm** [comment]
Hi @vobruba-martin ,

> I always use kickstart script to install netdata so I'm not familiar with manual installation. Is it enough to clone the branch, run cmake -DNETDATA_INTERNAL_CHECKS=1 and that's it? 

In normal situation, after to clone, I install Netdata with the following command:

```sh
 CFLAGS="-DNETDATA_INTERNAL_CHECKS=1" ./netdata-installer.sh
```

> Files in /etc/netdata and cache directory directories stay intact and after testing I can safely revert to the current version?

The installation won't remove any file.

If you prefer, we already merged the PR, so tomorrow the solution will be available with our `kickstart`. As soon I test it, I will come here to give you an update and you will be able to use the `kickstart`.

**Comment by vobruba-martin** [comment]
> If you prefer, we already merged the PR, so tomorrow the solution will be available with our `kickstart`. As soon I test it, I will come here to give you an update and you will be able to use the `kickstart`.

Yes, I will wait for this. Thank you!

**Comment by thiagoftsm** [comment]
Hello @vobruba-martin ,

I tested now and the `kickstart` already has the latest version of `eBPF.plugin` for `nightly`, if you run the next command:

```sh
wget -O /tmp/netdata-kickstart.sh https://my-netdata.io/kickstart.sh && sh /tmp/netdata-kickstart.sh
```

your system will be updated with it.

Best regards!

**Comment by vobruba-martin** [comment]
Hi @thiagoftsm,

Now I'm running `v1.33.1-42-nightly` with `socket = yes` and I can see that eBPF - network connections charts are back again. I will let you know after 1-2 days how is it going.

**Comment by thiagoftsm** [comment]
Hello @vobruba-martin ,

Thank you for your initial report! :) 
We will waiting for your feedback!

Best regards!

**Comment by vobruba-martin** [comment]
Sadly it didn't take long and the error is back :(

```
kernel: BUG: unable to handle page fault for address: ffffffffff600a1b
kernel: #PF: supervisor read access in kernel mode
kernel: #PF: error_code(0x0000) - not-present page
kernel: PGD 108e414067 P4D 108e415067 PUD 108e417067 PMD 108e419067 PTE 0
kernel: Oops: 0000 [#1] SMP NOPTI
kernel: CPU: 27 PID: 40431 Comm: mysqld Not tainted 5.13.0-28-generic #31~20.04.1-Ubuntu
kernel: Hardware name: HPE ProLiant DL380 Gen10 Plus/ProLiant DL380 Gen10 Plus, BIOS U46 11/03/2021
kernel: RIP: 0010:copy_from_kernel_nofault+0xbe/0x110
kernel: Code: 5d 5d c3 4c 89 e1 49 83 fd 03 76 16 31 c0 8b 13 89 11 85 c0 75 cf 48 83 c1 04 48 83 c3 04 49 83 ed 04 49 83 fd 01 76 18 31 c0 <66> 8b 13 66 89 11 85 c0 75 b1 48 83 c1 02 48 83 c3 02 49 83 ed 02
kernel: RSP: 0018:ff62421a4241fa60 EFLAGS: 00010246
kernel: RAX: 0000000000000000 RBX: ffffffffff600a1b RCX: ff62421a4241fb30
kernel: RDX: 0000000000000002 RSI: 0000000000000002 RDI: ffffffffff600a1b
kernel: RBP: ff62421a4241fa78 R08: 0000000000000001 R09: 0000000000000101
kernel: R10: 0000000000000000 R11: 0000000000000001 R12: ff62421a4241fb30
kernel: R13: 0000000000000002 R14: 0000000000000704 R15: 0000070400000000
kernel: FS:  00007eef4dffb700(0000) GS:ff431a6d2f780000(0000) knlGS:0000000000000000
kernel: CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
kernel: CR2: ffffffffff600a1b CR3: 000000307ff20001 CR4: 0000000000771ee0
kernel: DR0: 0000000000000000 DR1: 0000000000000000 DR2: 0000000000000000
kernel: DR3: 0000000000000000 DR6: 00000000fffe0ff0 DR7: 0000000000000400
kernel: PKRU: 55555554
kernel: Call Trace:
kernel:  <TASK>
kernel:  bpf_probe_read_compat+0x64/0x90
kernel:  bpf_prog_cc8225474521573b_netdata_tcp_sendmsg+0xce/0x6c4
kernel:  ? 0xffffffffc06e393c
kernel:  ? trace_call_bpf+0x5d/0xf0
kernel:  trace_call_bpf+0x66/0xf0
kernel:  kretprobe_perf_func+0x50/0x270
kernel:  ? tcp_schedule_loss_probe+0x115/0x1b0
kernel:  ? tcp_write_xmit+0x552/0xcc0
kernel:  kretprobe_dispatcher+0x6c/0x80
kernel:  ? inet_sendmsg+0x43/0x70
kernel:  __kretprobe_trampoline_handler+0xad/0xe0
kernel:  trampoline_handler+0x48/0x60
kernel:  kretprobe_trampoline+0x2a/0x50
kernel: RIP: 0010:kretprobe_trampoline+0x0/0x50
kernel: Code: 89 fc e8 a3 a5 01 00 4c 89 f2 4c 89 ee 4c 89 e7 44 0f b6 c0 31 c9 e8 ef 1c 36 00 41 5c 41 5d 41 5e 5d c3 cc cc cc cc cc cc cc <54> 9c 48 83 ec 18 57 56 52 51 50 41 50 41 51 41 52 41 53 53 55 41
kernel: RSP: 4241fde8:ff62421a4241fd88 EFLAGS: 00000286
kernel: RAX: 0000000000000704 RBX: 0000000000000040 RCX: 0000000000000000
kernel: RDX: 0000000000000000 RSI: ff431a3e96d1e008 RDI: ffffffffff600a0b
kernel: RBP: ff62421a4241fda0 R08: 0003a96417940f4a R09: 0000000000000101
kernel: R10: 0000000000000000 R11: 0000000000000001 R12: ff62421a4241fde8
kernel: R13: 0000000000000704 R14: ff431a6ecf16cec0 R15: 0000000000000000
kernel:  sock_sendmsg+0x5e/0x70
kernel:  __sys_sendto+0x113/0x190
kernel:  ? elfcorehdr_read+0x40/0x40
kernel:  ? exit_to_user_mode_prepare+0x3d/0x1c0
kernel:  ? ksys_read+0xce/0xe0
kernel:  __x64_sys_sendto+0x29/0x30
kernel:  do_syscall_64+0x61/0xb0
kernel:  ? irqentry_exit+0x19/0x30
kernel:  ? common_interrupt+0x55/0xa0
kernel:  ? asm_common_interrupt+0x8/0x40
kernel:  entry_SYSCALL_64_after_hwframe+0x44/0xae
kernel: RIP: 0033:0x7f175e71777c
kernel: Code: 89 4c 24 1c e8 85 f8 ff ff 44 8b 54 24 1c 8b 3c 24 45 31 c9 89 c5 48 8b 54 24 10 48 8b 74 24 08 45 31 c0 b8 2c 00 00 00 0f 05 <48> 3d 00 f0 ff ff 77 30 89 ef 48 89 04 24 e8 b1 f8 ff ff 48 8b 04
kernel: RSP: 002b:00007eef4dff91f0 EFLAGS: 00000246 ORIG_RAX: 000000000000002c
kernel: RAX: ffffffffffffffda RBX: 00007eef1c000b80 RCX: 00007f175e71777c
kernel: RDX: 0000000000000704 RSI: 00007eef1c01ee10 RDI: 0000000000000040
kernel: RBP: 0000000000000000 R08: 0000000000000000 R09: 0000000000000000
kernel: R10: 0000000000000040 R11: 0000000000000246 R12: 0000000000000704
kernel: R13: 0000000000000040 R14: 00007eef1c01ee10 R15: 0000000000000040
kernel:  </TASK>
kernel: Modules linked in: bonding tls tcp_bbr sch_fq intel_rapl_msr intel_rapl_common ipmi_ssif nfit x86_pkg_temp_thermal intel_powerclamp xfs coretemp crct10dif_pclmul ghash_clmulni_intel libcrc32c input_leds aesni_intel crypto_simd cryptd joydev rapl serio_raw mgag200 drm_kms_helper isst_if_mbox_pci cec rc_core isst_if_mmio i2c_algo_bit isst_if_common fb_sys_fops syscopyarea sysfillrect sysimgblt hpilo acpi_ipmi intel_pch_thermal ioatdma ipmi_si dca acpi_tad ipmi_devintf ipmi_msghandler mac_hid acpi_power_meter drm ip_tables x_tables autofs4 ses enclosure hid_generic xhci_pci usbhid smartpqi crc32_pclmul psmouse ice hid scsi_transport_sas xhci_pci_renesas intel_pmt wmi
kernel: CR2: ffffffffff600a1b
kernel: ---[ end trace 06e775e8e0c9acc3 ]---
kernel: RIP: 0010:copy_from_kernel_nofault+0xbe/0x110
kernel: Code: 5d 5d c3 4c 89 e1 49 83 fd 03 76 16 31 c0 8b 13 89 11 85 c0 75 cf 48 83 c1 04 48 83 c3 04 49 83 ed 04 49 83 fd 01 76 18 31 c0 <66> 8b 13 66 89 11 85 c0 75 b1 48 83 c1 02 48 83 c3 02 49 83 ed 02
kernel: RSP: 0018:ff62421a4241fa60 EFLAGS: 00010246
kernel: RAX: 0000000000000000 RBX: ffffffffff600a1b RCX: ff62421a4241fb30
kernel: RDX: 0000000000000002 RSI: 0000000000000002 RDI: ffffffffff600a1b
kernel: RBP: ff62421a4241fa78 R08: 0000000000000001 R09: 0000000000000101
kernel: R10: 0000000000000000 R11: 0000000000000001 R12: ff62421a4241fb30
kernel: R13: 0000000000000002 R14: 0000000000000704 R15: 0000070400000000
kernel: FS:  00007eef4dffb700(0000) GS:ff431a6d2f780000(0000) knlGS:0000000000000000
kernel: CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
kernel: CR2: ffffffffff600a1b CR3: 000000307ff20001 CR4: 0000000000771ee0
kernel: DR0: 0000000000000000 DR1: 0000000000000000 DR2: 0000000000000000
kernel: DR3: 0000000000000000 DR6: 00000000fffe0ff0 DR7: 0000000000000400
kernel: PKRU: 55555554
```

**Comment by thiagoftsm** [comment]
Hello @vobruba-martin ,

Thank you for your feedback. When I ran the tests, I could not recreate the problem, so I thought it had gone, but as you showed there is something that I am still missing. Tomorrow I will create another VM to recreate the problem.

**Comment by vobruba-martin** [comment]
Hello @thiagoftsm,

Maybe it is irrelevant but here is my `/etc/sysctl.conf` in which I've done some changes against the default one:

```
net.core.default_qdisc = fq
net.core.netdev_budget = 300000
net.core.netdev_budget_usecs = 50000
net.core.netdev_max_backlog = 3000000
net.core.netdev_tstamp_prequeue = 0
net.core.optmem_max = 100000000
net.core.rmem_default = 100000000
net.core.rmem_max = 100000000
net.core.somaxconn = 65535
net.core.wmem_default = 100000000
net.core.wmem_max = 100000000

net.ipv4.icmp_ratelimit = 2000
net.ipv4.ip_local_port_range = 2000 65535
net.ipv4.tcp_congestion_control = bbr
net.ipv4.tcp_fin_timeout = 15
net.ipv4.tcp_keepalive_intvl = 15
net.ipv4.tcp_keepalive_probes = 5
net.ipv4.tcp_keepalive_time = 300
net.ipv4.tcp_max_syn_backlog = 16384
net.ipv4.tcp_max_tw_buckets = 1440000
net.ipv4.tcp_mem = 6188763 8251686 12377526
net.ipv4.tcp_rfc1337 = 1
net.ipv4.tcp_rmem = 8192 87380 6291456
net.ipv4.tcp_synack_retries = 2
net.ipv4.tcp_tw_recycle = 0
net.ipv4.tcp_tw_reuse = 0
net.ipv4.tcp_wmem = 8192 16384 4194304
net.ipv4.udp_mem = 6188763 8251686 12377526
net.ipv4.udp_rmem_min = 16384
net.ipv4.udp_wmem_min = 16384

net.netfilter.nf_conntrack_max = 262144

vm.dirty_background_bytes = 104857600
vm.dirty_bytes = 209715200
vm.dirty_expire_centisecs = 3000
vm.dirty_writeback_centisecs = 500
vm.min_free_kbytes=1048510
vm.swappiness = 0
vm.vfs_cache_pressure = 100
vm.zone_reclaim_mode = 1
```

**Comment by thiagoftsm** [comment]
Thank you very much @vobruba-martin !!!
You gave us an important contribution, because I will have condition to be nearest your environment now.

**Comment by ktsaou** [comment]
@thiagoftsm I think the problem is related to the traffic itself. From the kernel OOPS, I see that the kernel is saying:

```
unable to handle page fault for address
```

This means that the kernel is trying to allocate some memory but it cannot for a reason. It does not necessarily mean it runs out of memory. It may mean that it cannot handle this memory allocation at that specific point.

On your lab, I suggest to setup something that has a lot of small packets transferred between 2 nodes, while the system is hit with new inbound sockets all the time. It is a timing issue, and if you are lucky you may be able to replicate it.

Probably speaking to the kernel devs at ubuntu (as @ThomasLamprecht suggested) is a better strategy. They may be able to tell what is happening.

**Comment by vobruba-martin** [comment]
Network data from the MySQL server during the error:
New connections: ~1000/s
Active connections: ~1500
Traffic IN: ~70 Mbit/s
Traffic OUT: ~300 Mbit/s

**Comment by thiagoftsm** [comment]
> @thiagoftsm I think the problem is related to the traffic itself. From the kernel OOPS, I see that the kernel is saying:

I agree, I am using Ubuntu to debug this. The configuration given [here](https://github.com/netdata/netdata/issues/12027#issuecomment-1048572541) is helping me to recreate some of the problems we are discussing in this issue.

**Comment by ktsaou** [comment]
@thiagoftsm I don't think we need mariadb for this. It happens also with ceph, so probably it should be happening with any other solution you could use to produce big volumes of new connections, active connections, traffic in and out.

**Comment by thiagoftsm** [comment]
> @thiagoftsm I don't think we need mariadb for this. It happens also with ceph, so probably it should be happening with any other solution you could use to produce big volumes of new connections, active connections, traffic in and out.

I said MariaDB, because I am years to work with MySQL. I expect they are more or less in the same level, with a small advantage for MariaDB.

I agree that a big sample of connections should be enough to replicate.

**Comment by thiagoftsm** [comment]
Hello @vobruba-martin,

Today I got the most poweful computer I have at my office, and I observed that when I run with default configuration, eBPF.plugin and the VM do not crash, on the other hand when I set the `sysctl.conf` that you gave me the VM crashes. Considering that extremes did not help me to detect which configuration is generating the problems and conducting the kernel for the `bug` situaton, I am adding configurations the configuration individually to `/etc/sysctl.conf` to identify exactly what is the configuration creating problems. I am also testing the option with different values to identify the range responsible to create the problem. 

I will continue with tests tomorrow.

Have a nice week!

**Comment by thiagoftsm** [comment]
I am reopening the issue, because final code was not merged in this repo.

**Comment by vobruba-martin** [comment]
@thiagoftsm 24 hours ago I updated Netdata to 1.34 and since then I'm going with `socket=yes`. I've got no errors so far. Thank you!

**Comment by thiagoftsm** [comment]
Thank you for your update @vobruba-martin !

We apologize for all problems we have created.

Best regards!
